SimpleTotp/TotpProvider.cs

```diff
@@ -0,0 +1,187 @@
+            return this.GetCodeAtSpecificTime(secretKey, time);
+        }
+
+        /// <inheritdoc />
+        public bool ValidateCode(String secretKey, String code, DateTimeOffset time)
+            => this.ValidateCode(secretKey, code, time, DefaultTolerance);
+
+        /// <inheritdoc />
+        public bool ValidateCode(String secretKey, String code, DateTimeOffset time, TimeSpan tolerance)
+            => this.ValidateCode(secretKey, code, time, tolerance, tolerance);
+
+        /// <inheritdoc />
+        public bool ValidateCode(String secretKey, String code, DateTimeOffset time, TimeSpan pastTolerance, TimeSpan futureTolerance)
+        {
+            ValidationHelper.CheckNotNullOrWhitespace(() => secretKey);
+            ValidationHelper.CheckNotNullOrWhitespace(() => code);
+
+            var validCodes = this.GetValidCodesForPeriod(secretKey, time, pastTolerance, futureTolerance);
+            return validCodes.Contains(code, StringComparer.OrdinalIgnoreCase);
+        }
+
+        /// <summary>
+        /// Returns a collection of valid codes between (<paramref name="time"/> - <paramref name="pastTolerance"/>)
+        /// and (<paramref name="time"/> + <paramref name="futureTolerance"/>).
+        /// </summary>
+        /// <param name="secretKey">User's secret TOTP key</param>
+        /// <param name="time">Reference point in time for specifying the valid time period (usually: current time)</param>
+        /// <param name="pastTolerance">Specified how far to the past will the valid period reach</param>
+        /// <param name="futureTolerance">Specified how far to the future will the valid period reach</param>
+        /// <returns>A collection of valid TOTP codes</returns>
+        protected internal IEnumerable<String> GetValidCodesForPeriod(String secretKey,
+                                                                      DateTimeOffset time,
+                                                                      TimeSpan pastTolerance,
+                                                                      TimeSpan futureTolerance)
+        {
+            ValidationHelper.CheckNotNullOrWhitespace(() => secretKey);
+
+            var minCounter = this.CalculateCounter(time - pastTolerance);
+            var maxCounter = this.CalculateCounter(time + futureTolerance);
+            var secretKeyBytes = Encoding.UTF8.GetBytes(secretKey);
+
+            for (var step = minCounter; step <= maxCounter; step++)
+            {
+                yield return this.CalculateCodeInternal(secretKeyBytes, step);
+            }
+        }
+
+        /// <summary>
+        /// Gets a TOTP code based on a secret and a counter
+        /// </summary>
+        /// <param name="secretKeyBytes">User's secret TOTP key as byte array</param>
+        /// <param name="counter">TOTP counter</param>
+        /// <returns>TOTP code</returns>
+        protected string CalculateCodeInternal(byte[] secretKeyBytes, long counter)
```

Comment:
Why this is called "Internal"? Is there "External" method also?

Reply:
done